### PR TITLE
User login and authentication.

### PR DESCRIPTION
**Pull request by atramors**
### Story :
[Add basic authentication to our API](https://trello.com/c/GBpRNjAZ/36-add-basic-authentication-to-our-api)

### Description : 

- Passwords inside of our DB now **hashed** (**POST** and **PUT** were updated):

```diff
+ hashed_password = generate_password_hash(kwargs["password"], method="sha256")
+ kwargs["password"] = hashed_password
```
- Changed constraints for user table:

```diff
- password = db.Column(db.String(60), nullable=False)
+ password = db.Column(db.String(80), nullable=False)
```
- **Login** part to get a token was made.
- **Token decorator** was created and added to endpoints 
Screenshot 2020-12-09 at 20 53 52

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,4 @@ mlruns/*
 carapp/__pycache__/*
 .DS_Store
 carapp/site.db
+carapp/.env
```
Comment (I159) [inline .gitignore:6]: Yeah, good practice ;)

File: carapp/__init__.py [hunk 2]
```diff
@@ -24,11 +29,9 @@
 Database migrations.
 """
 
-migrate = Migrate(app, db)
+migrate = Migrate(app, db, compare_type=True)
```
Comment (I159) [inline carapp/__init__.py:32]: Overall it is better to have a management command to migrate. It is out of the PR scope, so let's have a separate ticket for the improvement.

File: carapp/endpoints.py [hunk 3]
```diff
@@ -42,8 +48,11 @@ def put(self, user_id, **kwargs):
         logger.info(f"\nUser with id={user_id} updated")
         return result, 200
 
+    @token_required
     @use_kwargs(schemes.USER_ARGS)
     def post(self, **kwargs):
+        hashed_password = generate_password_hash(kwargs["password"], method="sha256")
```
Comment (I159) [inline carapp/endpoints.py:54]: Nice one!

File: carapp/security.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import jwt
+from werkzeug.security import check_password_hash
+from carapp import app, db_models
+from flask import make_response, request
+from functools import wraps
+
+# app.config["SECRET_KEY"] = "thisissecret"
+
+
+def token_required(any_function):
```
Comment (I159) [inline carapp/security.py:11]: This one is cool. And I believe was cognitive :) Let's have a separate ticket to use a third-party library for the implementation of the Bearer auth.

**Review by I159 — APPROVED** [review]
Good work! PR produces several new tickets. The auth should be evolved a little but the PR itself is completed.

File: carapp/__init__.py (hunk 2 above)
Comment (atramors) [reply carapp/__init__.py:32]: Yes sure. I have used command to migrate but it was necessary to add this change(**compare_type=True**) to update password credentials in DB (unfortunately, flask-migration can't catch such changes without this manipulation).

File: carapp/endpoints.py (hunk 3 above)
Comment (atramors) [reply carapp/endpoints.py:54]: Thanks!

File: carapp/security.py (hunk 4 above)
Comment (atramors) [reply carapp/security.py:11]: Yes sure, cool!

**Pull request merged**